### PR TITLE
test(gamification): Playwright e2e suite for HU-GAM-01, 08, 06

### LINKED ISSUE #2: Test: User can view notifications
# Test: User can view notifications

**Domain:** Core

## User Story

As a user, I want to view notifications so that see their notifications in real-time.

## Descripción detallada

Users want to see their notifications in real-time 
This story focuses on the core functionality. Advanced features, edge cases, and optimizations may be addressed in separate stories.

## Acceptance Criteria

- [ ] User can access the view
- [ ] All relevant information is displayed
- [ ] Data is formatted correctly
- [ ] Loading states are shown while fetching data
- [ ] Error messages are clear and actionable

### Backend
- GET endpoint for resource retrieval
- Query parameters for filtering/pagination
- Response formatting and serialization

### Frontend
- Component for main UI
- Form validation (if applicable)
- Loading and error states
- Success/error notifications
- Responsive design for mobile and desktop
- API integration with backend endpoints

## Testing Notes

- Test with valid data (happy path)
- Test with invalid/missing required data
- Test with edge cases (empty strings, special characters, very long inputs)
- Test error handling and error messages
- Test loading states and async behavior
- Test with authenticated and unauthenticated users

## Test Data Required

- Test user accounts (authenticated ITESO users)
- Valid input data for all required fields
- Invalid input data for validation testing
- Edge case data (boundary values, special characters)

## Potential Mocks

- Email service (SMTP)

## Dependencies & Duplication Check

**Dependencies:** None identified

**Possible duplicates:** None

## Assumptions

- User is authenticated as an ITESO community member
- Backend API follows RESTful conventions
- Frontend uses Next.js and TypeScript
- Database is PostgreSQL
- Priority level was not specified

## ⚠️ Warnings

- Acceptance criterion may be too vague: 'Data is formatted correctly'

### LINKED ISSUE #3: Test: User can view their points balance
# Test: User can view their points balance

**Domain:** Gamification

## User Story

As a user, I want to view their points balance so that see how many points they have accumulated from sustainable actions.

## Descripción detallada

Users want to see how many points they have accumulated from sustainable actions 
This story focuses on the core functionality. Advanced features, edge cases, and optimizations may be addressed in separate stories.

## Acceptance Criteria

- [ ] User can access the view
- [ ] All relevant information is displayed
- [ ] Data is formatted correctly
- [ ] Loading states are shown while fetching data
- [ ] Error messages are clear and actionable

### Backend
- GET endpoint for resource retrieval
- Query parameters for filtering/pagination
- Response formatting and serialization

### Frontend
- Component for main UI
- Form validation (if applicable)
- Loading and error states
- Success/error notifications
- Responsive design for mobile and desktop
- API integration with backend endpoints

## Testing Notes

- Test with valid data (happy path)
- Test with invalid/missing required data
- Test with edge cases (empty strings, special characters, very long inputs)
- Test error handling and error messages
- Test loading states and async behavior
- Test with authenticated and unauthenticated users

## Test Data Required

- Test user accounts (authenticated ITESO users)
- Valid input data for all required fields
- Invalid input data for validation testing
- Edge case data (boundary values, special characters)

## Potential Mocks

- Database queries for unit tests

## Dependencies & Duplication Check

**Dependencies:** None identified

**Possible duplicates:** None

## Assumptions

- User is authenticated as an ITESO community member
- Backend API follows RESTful conventions
- Frontend uses Next.js and TypeScript
- Database is PostgreSQL
- Priority level was not specified

## ⚠️ Warnings

- Acceptance criterion may be too vague: 'Data is formatted correctly'

### PR DESCRIPTION
**Pull request by antoniopelayo1902**
## 📌 Summary

Cross-team testing cruzado del equipo Core sobre el módulo de Gamificación. Cubre 3 historias con Playwright E2E:

- **HU-GAM-01** view-points (6 casos)
- **HU-GAM-08** level-progression (7 casos, incluye boundary off-by-one en 100 pts)
- **HU-GAM-06** eco-impact (7 casos)

24/24 tests verdes en ~8s. **2 bugs encontrados, ambos arreglados.**

---

## 🔍 Type of Change

- [ ] Feature (new functionality)
- [x] Bug fix
- [ ] Agent (development tooling)
- [ ] Documentation
- [ ] Refactor / cleanup
- [ ] CI / DevOps
- [x] Testing

---

## 🧩 Scope

- [ ] Backend
- [x] Frontend
- [ ] Agents
- [ ] Docs
- [ ] CI / Infrastructure

---

## 🤖 Agent Details (if applicable)

Tests generados con Playwright MCP + Kiro IDE.

---

## 🧪 Testing

- [x] Local execution
- [x] Manual testing
- [x] Unit tests
- [ ] Not applicable

**Cobertura:**
- HU-GAM-01: 6 casos (happy path, consistencia UI/API, /profile + /dashboard, no-session, error 500, localización es-MX)
- HU-GAM-08: 7 casos (beginner, **100 pts off-by-one**, progress bar bounds, champion, max level, UI⇔API, normalización de negativos)
- HU-GAM-06: 7 casos (zero state, UI⇔API, sufijo kg, community avg ≥ 0, singular/plural, redondeo 2 decimales, error 500)

**Total:** 24 tests ✅ (20 casos de negocio + 4 setups de auth)

**Cómo correr los tests:**
```bash
# Requiere backend :8000 y frontend :3000 corriendo
# Usuarios seed con 0/100/300/600 pts
cd frontend
npm install
npm run test:e2e      # headless
npm run test:e2e:ui   # modo interactivo
```

---

## 📎 Related Context

- Dinámica de rotación fase final: equipo Core prueba código del equipo Gamificación
- Nuevo código de tests: `frontend/tests/gamification/` (setup + fixtures + 3 specs)
- Config: `frontend/playwright.config.ts`
- Dependencias: `@playwright/test` agregado a devDependencies
- Cambios mínimos en código de Gam:
  - `data-testid` en PointsBalance, FeaturedGamificationCard, EcoImpact — solo para selectores estables, sin cambios funcionales

---

## 🐛 Hallazgos

| ID  | Historia  | Severidad | Estado                              |
|-----|-----------|-----------|-------------------------------------|
| #1  | HU-GAM-01 | 🔴 Media  | ✅ Fix aplicado                     |
| #2  | HU-GAM-06 | 🟡 Baja   | ✅ Fix aplicado                     |
| Obs. #1 | HU-GAM-08 | 🟢    | Validación positiva (off-by-one 100 pts) |
| Obs. #2 | HU-GAM-08 | 🟢    | Validación positiva (normalización negativos) |

### 🔴 Bug #1 — HU-GAM-01: `PointsBalance` no consumía `/api/gamification/points/` (FIX APLICADO)

**Severidad:** Media (contrato de historia incumplido + código muerto + request duplicada en la misma página)
**Ubicación:** `frontend/src/components/gamification/PointsBalance.tsx`
**Hook huérfano (ahora en uso):** `frontend/src/hooks/useUserPoints.ts`

**Descripción:** La historia HU-GAM-01 especifica que la UI de "puntos acumulados" debe leer del endpoint `GET /api/gamification/points/`. Sin embargo, `PointsBalance` obtenía el valor desde `levelProgression.points` (vía `useLevelProgression`), y el hook `useUserPoints` —que sí consume el endpoint correcto— nunca se importaba en ningún lugar del frontend.

**Por qué es un bug (no una optimización):**
1. `PointsBalance` no muestra nivel ni progreso. Solo muestra el número total y los 3 movimientos más recientes. El argumento "hace una sola llamada a `/level-progression/` porque muestra ambas cosas" no aplica: el componente descartaba `current_level`, `next_level`, `progress_percent`, `points_to_next_level`, `is_max_level` — el 85% del payload.
2. Ya había otra llamada a `/level-progression/` en la misma página `/profile`, hecha por `FeaturedGamificationCard`. Resultado: dos llamadas al endpoint más pesado en la misma vista.
3. `useUserPoints` era código muerto (`grep -rn "useUserPoints" frontend/src/` devolvía solo la declaración).
4. Contrato de la historia incumplido.

**Consecuencias del comportamiento previo:**
- Código muerto en el frontend (hook `useUserPoints`).
- Si el backend cambiaba la shape de `/level-progression/`, `PointsBalance` se rompía por una razón no obvia.
- El endpoint `/points/` quedaba sin consumidores en el frontend.
- Request duplicada al endpoint pesado en `/profile`.

---

### 🟡 Bug #2 — HU-GAM-06: Tipo `UserImpact` con campo inexistente en backend (FIX APLICADO)

**Severidad:** Baja (no rompe UI, pero produce valores `undefined` silenciosos si alguien los usa)
**Ubicación:** `frontend/src/hooks/useUserImpact.tsx`

**Descripción:** El tipo TypeScript local declaraba `items_saved_from_waste: number`, pero `ImpactSerializer` en `backend/gamification/serializers/impact.py` solo expone: `items_reused`, `co2_avoided`, `community_average_items`, `community_average_co2`.

---

### 🟢 Observación #1 — Lógica de niveles: no hay off-by-one en 100 pts

**Severidad:** Ninguna (validación positiva)
**Ubicación:** `backend/gamification/services/level_progression.py`

Se añadió un test con un usuario de **exactamente 100 puntos** (`rodrigo@iteso.mx`) para descartar el off-by-one común. El servicio resuelve correctamente: `100 pts → Active Reuser` (no Beginner). La comparación `points >= level.min_points` en `_resolve_current_level` es correcta. Test `gam-08 #2` confirma.

---

### 🟢 Observación #2 — Normalización de puntos negativos

**Severidad:** Ninguna (comportamiento correcto)
**Ubicación:** `backend/gamification/services/level_progression.py:49` (`safe_points = max(points, 0)`)

Si el backend recibiera `points < 0` por corrupción o migración, el servicio lo normaliza a 0 y la barra de progreso no queda negativa. Test `gam-08 #7` confirma que la UI tampoco renderiza anchos negativos.

---

## 🔧 Fixes Aplicados

Todos los fixes son cambios mínimos y reversibles (cada uno en su propio commit) si el equipo Gam lo prefiere en un PR separado.

### Fix #1 — `PointsBalance` ahora consume `/api/gamification/points/` (resuelve Bug #1)

**Archivo:** `frontend/src/components/gamification/PointsBalance.tsx`
**Commit:** `0cb3083`
**Tests ajustados:** `gam-01-view-points.spec.ts` (casos #2, #5, #6)

**Solución:** Cambiar el hook usado por `PointsBalance` de `useLevelProgression` a `useUserPoints`. No hay cambio funcional para el usuario — mismo valor mostrado, misma semántica de auto-refresh (`useUserPoints` también honra el evento `reuse:points-updated` y el polling de 10s).

```diff
- import { useLevelProgression } from '@/hooks/useLevelProgression';
+ import { useUserPoints } from '@/hooks/useUserPoints';
...
-  const { levelProgression, isLoading, error, refetch } = useLevelProgression(
-    isAuthenticated, refreshTrigger,
-  );
+  const { points, isLoading, error, refetch } = useUserPoints(
+    isAuthenticated, refreshTrigger,
+  );
...
-  {(levelProgression.points || 0).toLocaleString('es-MX')}
+  {(points ?? 0).toLocaleString('es-MX')}
```

**Verificación:**
- Test `gam-01 #2` ahora asserta con `waitForResponse` que la request a `/api/gamification/points/` efectivamente ocurre y que el valor en el DOM coincide con el response del endpoint.
- `useUserPoints` deja de ser código muerto.
- Se elimina la doble llamada a `/level-progression/` en `/profile` (ahora solo `FeaturedGamificationCard` la hace).

---

### Fix #2 — `useUserImpact` type: campo inexistente en backend (resuelve Bug #2)

**Archivo:** `frontend/src/hooks/useUserImpact.tsx`
**Commit:** `a00d2b2`

**Solución:** Eliminar la línea del tipo.

```diff
 type UserImpact = {
   items_reused: number;
-  items_saved_from_waste: number;
   co2_avoided: number;
   community_average_items: number;
   community_average_co2: number;
 };
```

**Verificación:** Test `gam-06-eco-impact.spec.ts #2` (UI⇔API) compara cada campo devuelto por `/api/gamification/impact/` contra el DOM. Pasa con los 4 campos reales del serializer.

---

### Fix #3 — Duplicate identifier `ChallengeType` en `types/gamification.ts` (drive-by fix para desbloquear CI)

**Archivo:** `frontend/src/types/gamification.ts`
**Commit:** `7a52f73`

**Problema:** El archivo tenía tres tipos declarados dos veces (`ChallengeType`, `Challenge`, `UserChallenge` en líneas 62 y 112). TypeScript reportaba `error TS2300: Duplicate identifier 'ChallengeType'.` Esto bloqueaba **el CI de todos los PRs** abiertos contra `dev`. Confirmado como pre-existing en `origin/dev` — el último build del branch `dev` (commit `7a9ef40`, 14 abril) también falla por esta razón.

**Solución:** Eliminar el bloque duplicado (segunda aparición, ~31 líneas). Se conserva la primera declaración, que tiene una interfaz `UserChallenge` más completa con los campos `reward_claimed` y `reward_claimed_at`.

**Verificación:**
- `npx tsc --noEmit` → 0 errores
- `npx prettier --check` → formato válido
- CI Frontend Checks → ✅ pass (antes: ❌ failing)
- Ningún consumidor del archivo se rompe (`grep -rn "ChallengeType\|UserChallenge" src/` confirma que `ChallengesBoard.tsx`, `ChallengesBoard.test.tsx` y `GlobalChallengeToasts.tsx` siguen compilando).

**Por qué drive-by y no PR separado:** la regla del repo (CI Gate) bloquea el merge hasta que CI esté verde. El fix es una eliminación de 31 líneas que TypeScript considera duplicadas — impacto 0 en runtime. Documentado en commit separado para que sea trivial hacer `git revert` si se prefiere moverlo a otro PR.

---

### Fix #4 — Prettier formatting en `EcoImpact.tsx`

**Archivo:** `frontend/src/components/gamification/EcoImpact.tsx`
**Commit:** `c163c97`

**Problema:** Al agregar `data-testid` a un `<p>` rompí el límite de ancho de línea que Prettier requiere.

**Solución:** `npx prettier --write` compactó los atributos en una línea. Sin cambios funcionales.

---

## ✅ Checklist

- [x] Code builds and runs locally
- [x] Changes are small and focused
- [x] Code follows agreed standards
- [x] Documentation was updated if needed
- [x] No unrelated changes are included
- [x] AI-generated content was reviewed and understood

---

## 📝 Notes for Reviewers

**Out of scope:**
- No se modificó backend
- No se commitearon migraciones ni tokens JWT

**Fixtures de auth:** Los archivos JSON con tokens JWT (`tests/gamification/fixtures/.auth/*.json`) se generan localmente en cada corrida — NO están en git (cubiertos por `.gitignore`).

**Commits en esta branch:**
```
342e144 chore: revert BUGS.md gitignore rule
0cb3083 fix(gamification): PointsBalance now consumes /api/gamification/points/
7a52f73 fix(types): remove duplicate ChallengeType, Challenge, UserChallenge definitions
c163c97 style(gamification): apply Prettier to EcoImpact.tsx
a00d2b2 test(gamification): add Playwright e2e suite for HU-GAM-01, 08, 06
```

**CI:** 3/3 ✅ (Backend Checks · Frontend Checks · CI Gate)

### CODE REVIEW

**Comment by Ferreirafc1133** [comment]
esta bien, cachaste los dos bugs y los fixes son limpios.

nota: chocas con #162 en types/gamification.ts. tu version esta bien, la suya no. mergeate primero.

3 nits inline.

File: frontend/tests/gamification/fixtures/auth.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import { request, expect } from '@playwright/test';
+import path from 'path';
+import fs from 'fs';
+
+const API_BASE = 'http://127.0.0.1:8000/api';
```
Comment (Ferreirafc1133) [inline frontend/tests/gamification/fixtures/auth.ts:5]: API_BASE hardcoded, metelo a env.

File: frontend/tests/gamification/gam-01-view-points.spec.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
+import { test, expect } from '@playwright/test';
+import { storageStatePath } from './fixtures/auth';
+
+/**
+ * HU-GAM-01 — View Points
+ * Endpoint: GET /api/gamification/points/ => { points: N }
+ * UI: PointsBalance component (/profile, /dashboard)
+ *
+ * Note: PointsBalance renders the number from `levelProgression.points`
```
Comment (Ferreirafc1133) [inline frontend/tests/gamification/gam-01-view-points.spec.ts:9]: jsdoc stale.

File: frontend/.gitignore [hunk 3]
```diff
@@ -56,3 +56,9 @@ yarn-error.log*
 # typescript
 *.tsbuildinfo
 next-env.d.ts
+
+# Playwright e2e artifacts
```
Comment (Ferreirafc1133) [inline frontend/.gitignore:60]: esta linea no ignora nada, borrala.
